libkbfs/rekey_queue.go

```diff
@@ -37,148 +60,114 @@ var _ RekeyQueue = (*RekeyQueueStandard)(nil)
 func NewRekeyQueueStandard(config Config) *RekeyQueueStandard {
 	log := config.MakeLogger("RQ")
 	rkq := &RekeyQueueStandard{
-		config: config,
-		log:    log,
+		config:   config,
+		log:      log,
+		queue:    make(chan rekeyQueueEntry, rekeyQueueSize),
+		pendings: make(map[tlf.ID]<-chan error),
 	}
 	return rkq
 }
 
+func (rkq *RekeyQueueStandard) doneLocked(entry rekeyQueueEntry, err error) {
+	entry.ch <- err
+	close(entry.ch)
+	delete(rkq.pendings, entry.id)
+	rkq.wg.Done()
+}
+
+func (rkq *RekeyQueueStandard) done(entry rekeyQueueEntry, err error) {
+	rkq.lock.Lock()
+	defer rkq.lock.Unlock()
+	rkq.doneLocked(entry, err)
+}
+
+func (rkq *RekeyQueueStandard) work(ctx context.Context) {
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case entry := <-rkq.queue:
+			newCtx := ctxWithRandomIDReplayable(ctx, CtxRekeyIDKey,
+				CtxRekeyOpID, nil)
+			rkq.log.CDebugf(newCtx, "Processing rekey for %s", entry.id)
+			err := rkq.config.KBFSOps().Rekey(newCtx, entry.id)
+			rkq.done(entry, err)
```

Comment:
Maybe this case should be wrapped in a `func()`, with this `done` call being a `defer` so that `wg` still gets decremented even on a panic?